packages/protocol/src/bdx/BdxMessenger.ts

```diff
@@ -0,0 +1,287 @@
+    async readBlockAckEof(): Promise<BdxBlockAckEof> {
+        const response = await this.nextMessage([BdxMessageType.BlockAckEof]);
+        BdxMessage.assert(response, BdxMessageType.BlockAckEof);
+        return response.message;
+    }
+
+    /** Reads the next BlockAck message and returns the decoded message. */
+    async readBlockAck(): Promise<BdxBlockAck> {
+        const response = await this.nextMessage([BdxMessageType.BlockAck]);
+        BdxMessage.assert(response, BdxMessageType.BlockAck);
+        return response.message;
+    }
+
+    /** Sends a Bdx Error StatusReport message with the given protocol status. */
+    sendError(code: BdxStatusCode) {
+        return this.#sendStatusReport(GeneralStatusCode.Failure, code);
+    }
+
+    /** Encodes and sends a Bdx StatusReport message with the given general and protocol status. */
+    async #sendStatusReport(generalStatus: GeneralStatusCode, protocolStatus: BdxStatusCode, requiresAck?: boolean) {
+        await this.#exchange.send(
+            SecureMessageType.StatusReport,
+            BdxStatusMessage.encode({
+                generalStatus,
+                protocolStatus,
+            }),
+            {
+                requiresAck,
+                logContext: {
+                    generalStatus: GeneralStatusCode[generalStatus] ?? Diagnostic.hex(generalStatus),
+                    protocolStatus: BdxStatusCode[protocolStatus] ?? Diagnostic.hex(protocolStatus),
+                },
+            },
+        );
+    }
+
+    protected throwIfErrorStatusReport(message: Message, logHint?: string) {
+        const {
+            payloadHeader: { messageType },
+            payload,
+        } = message;
+        if (messageType !== SecureMessageType.StatusReport) return;
+
+        const { generalStatus, protocolId, protocolStatus } = BdxStatusMessage.decode(payload);
+        if (generalStatus !== GeneralStatusCode.Success) {
+            throw new BdxStatusResponseError(
+                `Received general error status for protocol ${protocolId}${logHint ? ` (${logHint})` : ""}`,
+                generalStatus,
+                protocolStatus,
+            );
+        }
+        if (protocolStatus !== BdxStatusCode.Success) {
+            throw new BdxStatusResponseError(
+                `Received general success status, but protocol status is not Success${logHint ? ` (${logHint})` : ""}`,
+                generalStatus,
+                protocolStatus,
+            );
+        }
+    }
+
+    close() {
+        return this.#exchange.close();
+    }
+
+    /**
+     * Ensure that the value is a safe number and that it is not too large. Matter spec allows also 64bit values, but they
+     * make little sense for now, so make sure we handle them as too large. MAX_SAFE_INTEGER is 2^53-1 and is
+     * enough for now.
+     */
+    asSafeNumber(value: number | bigint | undefined, context = "", bdxErrorCode = BdxStatusCode.Unknown): number {
```

Comment:
Maybe `asSafeInteger` for naming as there is no notion of "safe" otherwise.  Or could be "asByteLength" or something as that's what it's used for.

I would also make it static as I don't think there's any references to local fields.  I think this confused me before because I assumed it was used locally